netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -105,6 +128,59 @@ public void onGoAwayReceived(int lastStreamId, long errorCode, ByteBuf debugData
     });
   }
 
+  /**
+   * Sends a ping to the server.
+   */
+  void ping(final PingCallback callback, Executor executor) {
+    checkState(ctx != null);
+    Http2Ping newPing = null;
+    Stopwatch stopwatch = null;
+    ByteBuf buffer = null;
+    while (true) {
+      if (goAwayStatus != null) {
```

Comment:
Changed both Netty and OkHttp implementations to only fail the ping when the channel is closed.

If a go away status is available at that time, that is used to generate the cause of failure. Otherwise, it synthesizes an "unavailable" `StatusException`.